File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        # Init 
        fleet = 0
        car_fleet_time = float("-inf")
        
        # Cars which are ahead are checked first
        for pos, reach_time in sorted([(position[i], (target-position[i])/speed[i]) for i in range(len(position))], reverse=True):
            
            if reach_time > car_fleet_time:
                car_fleet_time = reach_time
                fleet += 1
                
        return fleet

File: 0853-car-fleet/0853-car-fleet.java
import java.util.Arrays;
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int fleet = 0;
        double carFleetTime = Double.NEGATIVE_INFINITY;

        int n = position.length;
        Car[] cars = new Car[n];
        for (int i = 0; i < n; i++) {
            cars[i] = new Car(position[i], (double)(target - position[i]) / speed[i]);
        }
        Arrays.sort(cars, (a, b) -> Double.compare(b.position, a.position));

        for (Car car : cars) {
            double reachTime = car.reachTime;
            if (reachTime > carFleetTime) {
                carFleetTime = reachTime;
                fleet++;
            }
        }

        return fleet;
    }

    private static class Car {
        int position;
        double reachTime;

        Car(int position, double reachTime) {
            this.position = position;
            this.reachTime = reachTime;
        }
    }
}
